### PR TITLE
Riemannian Levenberg-Marquardt

### PR DESCRIPTION
**Pull request by mateuszbaran**
I've started working on #166, but it's very WIP right now. One issue I don't yet know how to solve is specifying basis for the Jacobian -- when we have i.e. `DefaultOrthonormalBasis` we can just put it in `get_vector`, but when we have `ProjectedOrthonormalBasis` then we need to `get_basis` first inside `step_solver!`. I don't yet have a good idea how to determine whether `get_basis` is needed or not -- we probably need a new function for that in `ManifoldsBase`?

### CODE REVIEW

**Comment by kellertuer** [comment]
I think when looking at it yesterday my main overall problem was, that Jacobian is something we do not yet have so much support for.
What would be your required function? I would always think you either have an implicit Basis that provides `get_vector` like your first or some kind of `CachedBasis`.

**Comment by mateuszbaran** [comment]
With Jacobians there are sort of two separate issues. First, the API used by Levenberg-Marquardt method and second, the API for something like AD Jacobians. For the second case we already have
```julia
jacobian(
    M_dom::AbstractManifold,
    M_codom::AbstractManifold,
    f,
    p,
    backend::TangentDiffBackend,
)
```
which works for fine when no `get_basis` is needed. I think we just need a variant that takes bases, like 
```julia
jacobian(
    M_dom::AbstractManifold,
    M_codom::AbstractManifold,
    f,
    p,
    basis_dom::AbstractBasis,
    basis_codom::AbstractBasis,
    backend::TangentDiffBackend,
)
```
For the first case I'd suggest a similar API of the form
```julia
jacF(
    M_dom::AbstractManifold,
    M_codom::AbstractManifold,
    p,
    basis_dom::AbstractBasis,
    basis_codom::AbstractBasis,
)
```
where `f` and `backend` would be internal like in the case of passing gradient functions to optimization algorithms.

**Comment by kellertuer** [comment]
But the TangentDiffBackend has a Basis does it not? It does, even for Dom and codom (arg&val)

https://juliamanifolds.github.io/Manifolds.jl/dev/features/differentiation.html#Manifolds.TangentDiffBackend

**Comment by kellertuer** [comment]
Jacobian could store the backend?

I am not yet sure what the best way to proceed here is, because I do not yet see all consequences of the design choice we are deciding on here are.

**Comment by mateuszbaran** [comment]
The Jacobian function that is called by optimization function would definitely have to store the backend if that's how it would like to calculate the Jacobian. I more or less see now how to design this, I will let you know when that's ready.

**Comment by kellertuer** [comment]
Two small remarks I just saw

1. structure wise until now we usually have a `plan/` file to store the problem (if it is a new one and the algorithm is not based on an existing one) and the options in a `xyz_pplan.jl` to have it split from the actual solver and it can be reused without including the solver
2. I saw you already implemented `get_cost`, for debug (and also stopping criteria) it might be a good idea to maybe also implement `get_gradient`?

**Comment by mateuszbaran** [comment]
> 1. structure wise until now we usually have a `plan/` file to store the problem (if it is a new one and the algorithm is not based on an existing one) and the options in a `xyz_pplan.jl` to have it split from the actual solver and it can be reused without including the solver

Yes, I will restructure it after the functionality is mostly ready -- it's slightly easier to develop in one file.

> 2\. I saw you already implemented `get_cost`, for debug (and also stopping criteria) it might be a good idea to maybe also implement `get_gradient`?

I've also implemented `get_gradient` :slightly_smiling_face: .

**Comment by kellertuer** [comment]
Oh then I missed that, I only looked at the changes here and with the missing coverage remarks in there it's hard to see. Cool!

File: src/solvers/LevenbergMarquardt.jl [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    initial_jacF::TJac;
+    stopping_criterion::StoppingCriterion=StopAfterIteration(100),
+    retraction_method::AbstractRetractionMethod=default_retraction_method(M),
+) where {P,TJac}
+    return LevenbergMarquardtOptions{P}(
+        M, x, initial_jacF, stopping_criterion, retraction_method
+    )
+end
+
+function get_gradient(o::LevenbergMarquardtOptions)
```
Comment (kellertuer) [inline src/solvers/LevenbergMarquardt.jl:172]: The `Options` `o` should not store the manifold. Note that 

https://manoptjl.org/stable/plans/problem/#Manopt.get_gradient

has two meanings (maybe we should change that):

1. with options it returns the last evaluated gradient (stored within `O`)
2. with `Problem` `P` and a point `x` it evaluates the gradient stored / computed from the static part in P (that is the manifold and the cost function and the gradient function and so on).

Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl:172]: also Jack seems to be similar to the gradient a static/unmodified function/thing that should be stored in the Problem?

Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:172]: Note that to implement `get_gradient` I need to either store gradient in options or compute it from what is stored there. And `step_solver!` doesn't explicitly compute a gradient, so I opted to store what is actually computed anyway (Jacobian of F) and compute gradient as needed.

Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:172]: > also Jack seems to be similar to the gradient a static/unmodified function/thing that should be stored in the Problem?

Could you rephrase that? I'm not sure what you mean.

Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:172]: > And step_solver! doesn't explicitly compute a gradient,
Or maybe it should, let me think about it.

Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl:172]: > Note that to implement `get_gradient` I need to either store gradient in options or compute it from what is stored there. And `step_solver!` doesn't explicitly compute a gradient, so I opted to store what is actually computed anyway (Jacobian of F) and compute gradient as needed.

The gradient should be able to be computed given the problem (containing the static data like the cost and all its derivatives/subgradiens/jacobians of inner vector stuff) and an iterate. You could use a function `get_jacobian` which you introduce and use in the solver, which could also be internally used by `get_gradient` then?

File: src/solvers/LevenbergMarquardt.jl [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    F::TF
+    jacobian!!::TJ
+end
+function NonlinearLeastSquaresProblem(
+    M::mT, F::TF, jacF::TJ; evaluation::AbstractEvaluationType=AllocatingEvaluation()
+) where {mT<:AbstractManifold,TF,TJ}
+    return NonlinearLeastSquaresProblem{typeof(evaluation),mT,TF,TJ}(M, F, jacF)
+end
+
+function (d::DebugGradient)(::NonlinearLeastSquaresProblem, o::Options, i::Int)
```
Comment (kellertuer) [inline src/solvers/LevenbergMarquardt.jl:88]: If you problem is an AbstractGradientProblem this might – for best of cases – work directly without an implementation?

Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:88]: I'm still unsure how to model it, and for now I'm focusing more on making it actually work. I will work out the structure when I'm sure it works, OK?

Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl:88]: That is of course totally fine, Then we keep this as a reminder. I can also check in the end where we can spare a few implementations by inheritance then.

**Comment by mateuszbaran** [comment]
@kellertuer how should I extend `get_last_stepsize` so that I can use `StopWhenStepsizeLess` with `LevenbergMarquardt`? BTW, the `vars...` thing in `get_last_stepsize` caused a very nasty stack overflow error, could it be perhaps replaced by more specific argument lists?

**Comment by mateuszbaran** [comment]
Also, it would be nice to be able to replicate the stopping criterion from this paper: https://publications.ics.forth.gr/_publications/0201-P0401-lourakis-levenberg.pdf (the one with $\varepsilon_2$), though it would be quite nontrivial in general.

File: src/solvers/LevenbergMarquardt.jl (hunk 1 above)
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl]: I've removed manifold from options, in the end it turned out to be not needed.

**Comment by kellertuer** [comment]
> @kellertuer how should I extend `get_last_stepsize` so that I can use `StopWhenStepsizeLess` with `LevenbergMarquardt`? BTW, the `vars...` thing in `get_last_stepsize` caused a very nasty stack overflow error, could it be perhaps replaced by more specific argument lists?

Can you be a little more precise what you would need for this? I had hoped the stepsize-Thingy would not be dependent on with which options it is used. Also the vars – feel free to improve that (until now it worked fine, but sure nasty things should not happen).

**Comment by kellertuer** [comment]
> Also, it would be nice to be able to replicate the stopping criterion from this paper: https://publications.ics.forth.gr/_publications/0201-P0401-lourakis-levenberg.pdf (the one with 
> ε
> 2
> ), though it would be quite nontrivial in general.

At first glance, I do not see what ``\delta p`` in their algorithm is, but we can surely also implement a new stopping criterion for that if that is not covered by the ones we currently have.

**Comment by mateuszbaran** [comment]
> Can you be a little more precise what you would need for this? I had hoped the stepsize-Thingy would not be dependent on with which options it is used. Also the vars – feel free to improve that (until now it worked fine, but sure nasty things should not happen).

Basically the idea is to terminate if `o.stepsize` is smaller than a given constant but currently `get_last_stepsize` looks very convoluted to me so I'm not sure how to modify it to accommodate this use case. I will try to clean up `get_last_stepsize` a bit.

> At first glance, I do not see what `\delta p` in their algorithm is

$\delta_p$ is their step, same thing as `step_vector` in my algorithm. The issue is with norm of `p`, which only makes sense on certain manifolds.

**Comment by kellertuer** [comment]
Ah, ok let's take a look

https://github.com/JuliaManifolds/Manopt.jl/blob/101b7a90ada52ec25536ad8f2143efafb8d30ca2/src/plans/stepsize.jl#L828-L846

it is a little crowded since it does two things
1) lines 828-836 “unpack” the options, since they might be inside `DebugOptions` or `RecordOptions`, that is also why that has `vars...` since it should just pass all remaining ones further on, for example usually the iterate is also passed down, since line 839 need that (for example for the decreasing stepsize and because the default interface for step sizes is a call with `(P,O,i)` parameters.
2) the 839-846 are specific checks where to get the stepwise from, where the default is line 839, we just call the stepwise functor to compute the stepsize if this (the `I`th) iteration again (works for things like constant step size or decreasing step size and such).
3) 8.40-846 are the special cases where the last step size is stored, then these just access this last step size in there instead of computing it again. 
840

**Comment by mateuszbaran** [comment]
Hm, OK, I think I see.

**Comment by kellertuer** [comment]
Sure that might be better documented, you are right in that,
but let me know if we should improve this (generally – sure – two layers would be nice for example).

**Comment by mateuszbaran** [comment]
Hm, it appears that in the test suite in `vars` we can only have an integer?

**Comment by kellertuer** [comment]
> Hm, it appears that in the test suite in `vars` we can only have an integer?

I would phrase that as: Until now we only used it to carry the integer (iteration number) over.

**Comment by kellertuer** [comment]
Please ignore the ± 10 lines in stepsize, I still have not yet found the time to fix that wiggle within WolfePowell, which has (sadly) a random-based test.

**Comment by mateuszbaran** [comment]
Sure, I didn't change anything around that so I assumed it was something random :slightly_smiling_face: . The PR is close to being ready for review, I just need to finish support for cached bases and improve documentation a bit.

**Comment by mateuszbaran** [comment]
I think this is now more or less ready for review.

**Review by kellertuer — COMMENTED** [review]
Code-coverage looks great! I have a few remarks, where the jacobian one we already discussed before but if possible I prefer access functions for problem-struct-stuff.

File: src/plans/nonlinear_least_squares_plan.jl [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+* `jacF` – the current Jacobian of ``F``
+* `gradient` – the current gradient of ``F``
+* `step_vector` – the tangent vector at `x` that is used to move to the next point
+* `last_stepsize` – length of `step_vector`
+* `η` – parameter of the algorithm, the higher it is the more likely the algorithm will be
+  to reject new proposal points
+* `damping_term` – current value of the damping term
+* `damping_term_min` – initial (and also minimal) value of the damping term
+* `β` – parameter by which the damping term is multiplied when the current new point is rejected
+* `flagnz` – (`true`) if false, the algorithm expects that the value of residual at mimimum
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:132]: Good idea -- I took the name from the paper but `expect_zero_residual` would be more clear.

File: src/solvers/LevenbergMarquardt.jl [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+Solve an optimization problem of the form
+
+```math
+\operatorname{minimize}_{x ∈ \mathcal M} \frac{1}{2} \lVert F(x) \rVert^2,
+```
+
+where ``F: \mathcal M \to ℝ^d`` is a continuously differentiable function,
+using the Riemannian Levenberg-Marquardt algorithm [^Peeters1993].
+The implementation follows [^Adachi2022].
+
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:13]: Hm, I don't think I can do a better job at explaining the algorithm than Adachi2022 (which is freely available so anyone can take a look).

File: src/solvers/LevenbergMarquardt.jl [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    > and Local Convergence Properties.” arXiv, Oct. 01, 2022.
+    > doi: [10.48550/arXiv.2210.00253](https://doi.org/10.48550/arXiv.2210.00253).
+[^Peeters1993]:
+    > R. L. M. Peeters, “On a Riemannian version of the Levenberg-Marquardt algorithm,”
+    > VU University Amsterdam, Faculty of Economics, Business Administration and Econometrics,
+    > Serie Research Memoranda 0011, 1993.
+    > link: [https://econpapers.repec.org/paper/vuawpaper/1993-11.htm](https://econpapers.repec.org/paper/vuawpaper/1993-11.htm).
+"""
+function LevenbergMarquardt(
+    M::AbstractManifold, F::TF, gradF::TDF, x, num_components::Int; kwargs...
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:47]: For allocating evaluation and at the cost of an additional evaluation, yes.

File: src/solvers/LevenbergMarquardt.jl [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    jacF::TDF,
+    x,
+    num_components::Int;
+    retraction_method::AbstractRetractionMethod=default_retraction_method(M),
+    stopping_criterion::StoppingCriterion=StopAfterIteration(200) |
+                                          StopWhenGradientNormLess(1e-12) |
+                                          StopWhenStepsizeLess(1e-12),
+    debug=[DebugWarnIfCostIncreases()],
+    evaluation::AbstractEvaluationType=AllocatingEvaluation(),
+    flagnz::Bool=true,
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:71]: This is the same thing as `flagnz` in `LevenbergMarquardtOptions`, I will change the name here as well.

File: src/plans/nonlinear_least_squares_plan.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:132]: I know that the name is from the paper, I did not like the name their, either ;)

File: src/solvers/LevenbergMarquardt.jl [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    return get_solver_return(solve(p, o))
+end
+#
+# Solver functions
+#
+function initialize_solver!(
+    p::NonlinearLeastSquaresProblem{AllocatingEvaluation}, o::LevenbergMarquardtOptions
+)
+    o.Fval = p.F(p.M, o.x)
+    o.gradient = get_gradient(p, o.x)
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:97]: Yes, this is also supposed to allocate memory here.

File: src/solvers/LevenbergMarquardt.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl:13]: I can check, I prefer to make the docs a little bit self-contained.

File: src/solvers/LevenbergMarquardt.jl [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    end
+end
+
+function step_solver!(
+    p::NonlinearLeastSquaresProblem{Teval}, o::LevenbergMarquardtOptions, iter::Integer
+) where {Teval<:AbstractEvaluationType}
+    # o.Fval is either initialized by initialize_solver! or taken from the previous iteraion
+
+    basis_ox = _maybe_get_basis(p.M, o.x, p.jacB)
+    if Teval === AllocatingEvaluation
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:122]: Good question, though here the meaning of Jacobian is tied to the nonlinear least squares problem so it doesn't seem as universal as `get_gradient`.

I will add documentation about `B_dom` somewhere.

File: src/solvers/LevenbergMarquardt.jl [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    sk = cholesky(JJ) \ -grad_f_c
+    get_vector!(p.M, o.gradient, o.x, grad_f_c, basis_ox)
+
+    get_vector!(p.M, o.step_vector, o.x, sk, basis_ox)
+    o.last_stepsize = norm(p.M, o.x, o.step_vector)
+    temp_x = retract(p.M, o.x, o.step_vector, o.retraction_method)
+
+    normFk2 = norm(o.Fval)^2
+    if Teval === AllocatingEvaluation
+        o.Fval_temp = p.F(p.M, temp_x)
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:142]: We don't have to store the temp, It's just to avoid memory allocations. I will rename it as you suggested.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl:142]: That is of course reasonable to avoid memory allocations then :)

File: src/plans/nonlinear_least_squares_plan.jl [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    F::TF
+    jacobian!!::TJ
+    jacB::TB
+    num_components::Int
+end
+function NonlinearLeastSquaresProblem(
+    M::mT,
+    F::TF,
+    jacF::TJ,
+    num_components::Int;
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:52]: I've fixed this.

File: src/solvers/LevenbergMarquardt.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl]: I've renamed this variable.

File: src/plans/nonlinear_least_squares_plan.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl]: I've changed the name.

File: src/solvers/LevenbergMarquardt.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl]: I've changed the name.

File: src/solvers/LevenbergMarquardt.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl]: I've resolved this.

File: src/solvers/LevenbergMarquardt.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:137]: I've chosen `B_dom` because in general to calculate a Jacobian we need bases for both domain and codomain, though here I've opted to assume that in the codomain the standard Euclidean basis is used. It's more about potential consistency with cases where more general Jacobians might be useful. Maybe it's not worth it though.

Regarding the other point, I'm not sure how `get_jacobian` would help simplify `step_solver!`?

File: src/plans/nonlinear_least_squares_plan.jl [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+end
+
+function get_cost(P::NonlinearLeastSquaresProblem{AllocatingEvaluation}, p)
+    return 1//2 * norm(P.F(P.M, p))^2
+end
+function get_cost(P::NonlinearLeastSquaresProblem{MutatingEvaluation}, p)
+    Fval = zeros(P.num_components)
+    P.F(P.M, Fval, p)
+    return 1//2 * norm(Fval)^2
+end
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: What exactly would `get_cost!` do?

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: `get_cost!(P, Fval, p)` would evaluate the (vectorial) cost in place, similarly we could use a `get_cost(P, p)` to allocate (which is done here anyways).

File: src/solvers/LevenbergMarquardt.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl:137]: `get_jacobian` would do the `if Teval === AllocatingEvaluation` via dispatch (as do all other functions accessing “stuff” from the problem).

Concerning the `B_dom` I would then maybe prefer `basis_domain`?

File: src/plans/nonlinear_least_squares_plan.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: I'm a little worried that it might be confusing as that's not really cost but the residual, with the cost being half of the norm of the residual.

Also note that to simplify
```julia
    if Teval === AllocatingEvaluation
        o.candidate_cost_values = p.F(p.M, temp_x)
    else
        p.F(p.M, o.candidate_cost_values, temp_x)
    end
```
we would have to pass entire options structure to such `get_cost`, which is not a pattern I've seen in Manopt so far.

File: src/solvers/LevenbergMarquardt.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/solvers/LevenbergMarquardt.jl:137]: OK, I will change the name to `basis_domain`.

Regarding `get_jacobian`, I think we have the same issue as in https://github.com/JuliaManifolds/Manopt.jl/pull/167#discussion_r1038007228 .

File: src/plans/nonlinear_least_squares_plan.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: Then let's call it `get_residuals`(I thought it was costs) - also surprising that it is then stored in something that is called costs if it is not costs (yes I proposed the name but for the same reason)

Concerning the passing – no? 

a) would be `o.candidate_cost_values = get_residuals(p, temp_x)`
b) would be `get_residuals!(p, o.candidate_cost_values, temp_x)`

and actually I would just use a) and have 4 variants of `get_residuals` / `get_residuals!` (two each, namely one for Allocating one for Mutating for each of the functions).

File: src/solvers/LevenbergMarquardt.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl:137]: Hm, again, maybe just “steal” from `get_gradient`?

File: src/plans/nonlinear_least_squares_plan.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: > Then let's call it `get_residuals`(I thought it was costs) - also surprising that it is then stored in something that is called costs if it is not costs (yes I proposed the name but for the same reason)

Maybe calling it costs wasn't a particularly good choice. Nonlinear least squares is primarily used for regression-type problems where `F` returns residuals. In particular it has nothing to do with multi-objective optimization.

> Concerning the passing – no?
> 
> a) would be `o.candidate_cost_values = get_residuals(p, temp_x)` b) would be `get_residuals!(p, o.candidate_cost_values, temp_x)`
> 
> and actually I would just use a) and have 4 variants of `get_residuals` / `get_residuals!` (two each, namely one for Allocating one for Mutating for each of the functions).

Yes, that would work but would
```julia
    if Teval === AllocatingEvaluation
        o.candidate_cost_values = get_residuals(p, temp_x)
    else
        get_residuals!(p, o.candidate_cost_values, temp_x)
    end
```
be much nicer than the current code?

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: That won't work if `o.candidate_cost_values` is for example `SArray`, which I think would be nice to support.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: hm, but then nearly all of my accessor function have this problem at the moment, we should talk about that – maybe even in a small meeting on zoom?

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: Because until now I thoght the approach with 4 functions – sure it might allocate for the allocating variant, but there is nothing one can do about that; I am referring to the 4 functions like

https://github.com/JuliaManifolds/Manopt.jl/blob/40ec55022f3aefc51ae28347c123621cbaa4cd02/src/plans/gradient_plan.jl#L54-L68

which encapsulate nicely whether its mutating or not – and the user can use this or that variant (`get_gradient`/ `get_gradient!`) to indicate basically “if this is possible, just mutate do not allocate” and the `!` variant for allocating functions – well still allocates and then copies.

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: > hm, but then nearly all of my accessor function have this problem at the moment, we should talk about that – maybe even in a small meeting on zoom?

It's not a big issue but it would be nice to use `SArray` for gradients, for example, in the allocating evaluation mode. We can schedule a zoom meeting on Slack if you want :slightly_smiling_face: .

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: In the latest commit I've sketched a possible alternative we've discussed. What do you think? I know `access_field(o, :jacF, Teval())` isn't particularly readable but there is not that much that can be done. Alternatively we could have `(@access_field o.jacF)` with `Teval` being implicitly captured.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: For me that does look ok, I just have not yet understood where we need the `FieldAccessor` or another words how complicated it is for the user to setup the `o.jacF` now ...

Ah, I think I got it, so basically for the allocating case, we create an `FieldAcessor` which does the job (maybe comparable to a view?) so

a) for the user it would enable to _just_ use it as before?
b) we internally need this small trick with `acess_field`? And alternatively we could pass `o` to the `get_jacobian(!)` and do the access there?

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: I think I would need these functions and types properly documented to understand the last few details but besides that the `access_field` seems a good solution and the macro looks even nicer if that would be possible, then we are nearly back to classical code from before.

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: Yes, the user sees no difference and people writing new methods just need to remember to use `access_field`. And `FieldAccessor` is actually more comparable to a reference (`Ref`). I will add some documentation and the macro.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: Ah yes, but also `Ref` us basically a “short-time-wrap”. Thanks for that idea, looking forward to using that. I will wait on the other branch for this to be finished, to directly merge and adopt it then :)

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/plans/nonlinear_least_squares_plan.jl:74]: Cool :slightly_smiling_face: . I've added some docs, macro and changed the type name to `FieldReference` to make it seem more related to `Ref`.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/plans/nonlinear_least_squares_plan.jl:74]: This looks quite cool – is nearly as easy as the old variant and with the macro still very nice within the algorithms! Thanks.

File: src/solvers/LevenbergMarquardt.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/solvers/LevenbergMarquardt.jl]: This is also resolved with the new variant.

**Comment by kellertuer** [comment]
It seems only the docs update (will hopefully find time tomorrow) and maybe a few tests is the only things missing?

**Comment by mateuszbaran** [comment]
Yes, I think that's all that needs to be done. I'm finishing the tests right now, and I will wait for your docs :slightly_smiling_face: .

**Comment by kellertuer** [comment]
Im, you are probably right, that algorithm is not so easy to summarise in a few lines. Could you maybe add “Algorithm 1” behind the Riemannian reference (the arXiv one)?

**Comment by kellertuer** [comment]
> Im, you are probably right, that algorithm is not so easy to summarise in a few lines. Could you maybe add “Algorithm 1” behind the Riemannian reference (the arXiv one)?

Fixed it myself. I think – if code coverage is fine (I think we currently define TestProblem twice in the tests) – then just a bump of the version is missing.

**Comment by kellertuer** [comment]
I have maybe one final question: Since I in general would not care what the types are, is it maybe also reasonable to do the “wrapping” in Manifolds? Maybe that is not possible because we have to know that it is a field of some struct? Otherwise one could think of doing that on the `copy!`level maybe?

**Comment by mateuszbaran** [comment]
Thanks for improving the reference. I will fix the problem with tests I've just introduced and bump version.

> I have maybe one final question: Since I in general would not care what the types are, is it maybe also reasonable to do the “wrapping” in Manifolds? Maybe that is not possible because we have to know that it is a field of some struct? Otherwise one could think of doing that on the `copy!`level maybe?

What do you mean by doing the wrapping in Manifolds.jl? We currently don't really need such wrapping but I think we could put `FieldReference` in ManifoldsBase.jl if that's what you mean.

**Comment by kellertuer** [comment]
> What do you mean by doing the wrapping in Manifolds.jl? We currently don't really need such wrapping but I think we could put `FieldReference` in ManifoldsBase.jl if that's what you mean.

Yes, that is what I mean, since it is a problem addressing the “assignment of a value”, that is basically a problem of `copyto!` (and not of `get_gradient!` or these), then the functions here could return to using `copyto!` – that was what I head in mind at least.

**Review by kellertuer — APPROVED** [review]

**Comment by mateuszbaran** [comment]
Hm, then that's not what I had in mind. `copyto!` isn't defined for references so I don't think it would be fine to make it work like that (it would be unclear whether it means "copy to referenced value" or "change referenced value").

**Comment by kellertuer** [comment]
hm but it is more general than just something I would use within evaluating a gradient as well...

**Comment by kellertuer** [comment]
Well, independent of that we can surely merge this here and think about introducing that in general in Manifolds (I think that would be nice) – thanks for introducing this algorithm to Manopt.jl! 👍 😎

**Comment by mateuszbaran** [comment]
Hm, I was thinking about moving `FieldReference` to ManifoldsBase.jl but it can stay here in Manopt.jl for now.

> thanks for introducing this algorithm to Manopt.jl!

You're welcome, I also learned quite a bit working on it. Now I'm wondering if Manopt.jl regression examples should use Levenberg-Marquardt? Regression is exactly what it was designed to handle.

**Comment by kellertuer** [comment]
Sure, the simple regression can surely also be done with LM,, I think (the fixed time point one).
In 0.4 I want to reduce the number of “dead” examples (that is the non-Pluto ones that are not run often), so if so – please to a Pluto one ;)
